### ISSUE TITLE
PERF401 doesn't handle defaultdicts

### ISSUE THREAD

**Issue by hynek** [astral-sh/ruff#5500]
Given this code:

```python
from collections import defaultdict

d = defaultdict(list)

for i in [1, 2, 3]:
    d[i].append(i**2)
```

Gives me this:

```python
$ pipx run ruff --version
ruff 0.0.276
$ pipx run ruff --isolated --select PERF401 t2.py
t2.py:6:5: PERF401 Use a list comprehension to create a transformed list
Found 1 error.
```

Maybe I'm dense, but I don't think I can rewrite that as a list comprehension, non the less because it's not a list, but a (default)dict. 😅

Interesting fact: the error doesn't trigger if there's no computation in `append()`.

**Comment by jerr0328** [comment]
Came here for similar reasons, though I'm getting PERF402, seems like #5494 is similar.

**Comment by charliermarsh** [comment]
I believe this is a dupe of #5494 - we need to avoid triggering when the loop variable is used on the left-hand side (or perhaps just avoid whenever the left-hand side is more complicated than a name).

**Comment by nijel** [comment]
I have no clue about the implementation,  but there are different checks being triggered in the issues PERF402 and PERF401.

**Comment by charliermarsh** [comment]
Yup acknowledged — it’s the same underlying problem and fix.
